ntp-udp/src/interface.rs

```diff
@@ -6,61 +6,125 @@
 //! Uses the Linux and/or BSD specific function `getifaddrs` to query the list
 //! of interfaces and their associated addresses.
 
-use std::ffi;
 use std::iter::Iterator;
-use std::mem;
 use std::net::SocketAddr;
 use std::option::Option;
 
-#[allow(dead_code)]
-pub fn interface_name(local_addr: SocketAddr) -> std::io::Result<Option<[libc::c_char; 16]>> {
-    let matches_inferface = |interface: &InterfaceAddress| match interface.address {
-        None => false,
-        Some(address) => address.ip() == local_addr.ip(),
-    };
+use crate::raw_socket::interface_iterator::InterfaceIterator;
 
-    if let Some(interface) = getifaddrs()?.find(matches_inferface) {
-        let mut ifrn_name = [0; 16];
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct InterfaceName {
+    bytes: [u8; libc::IFNAMSIZ],
```

Comment:
it is so that `as_str` and `as_cstr` work without any unsafe code. the mapping of the array from `u8` to `i8` (which is what `libc::c_char` turns out to be usually) is without runtime cost, but certain `std` functions want a `&[u8]` and we cannot cast to that from `&[libc::c_char]` without unsafe